### DISCUSSION TITLE
Welcome to kube-hetzner Discussions!

### DISCUSSION THREAD

**Discussion by mysticaltech** [kube-hetzner/terraform-hcloud-kube-hetzner#76] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by phaer** [comment]
Great initiative, thanks! :rocket:

**Reply by mysticaltech** [reply]
Thanks for pushing! :)

**Comment by nupplaphil** [comment]
Great to have this board!

I'm currently a docker enthusiast and I'm curious about using kubernetes and hetzner (since i'm currently heavily using the hetzner root server) in the future instead of "just" docker. I already created a ansible boilerplate for kubernetes and hetzner, but it wasn't as professional as this solution ;-)

Currently I'm not sure how to structure my new environment. One part will include this repo and extend it (first question is how to keep on track with the changes here and my customization). Another part is extending the infrastrucutre with things like the cert-manager, minio for s3 and similar things. And the third and last part will be the apps, I do want to manage (like Next cloud, mailserver, gitea, woodpecker,...).

How are you managing these three parts at your setup? Using three different root paths? Combined at one structure based on this repo?

What about the database? Do you recommend using Cassandra with your kubernetes solution or rely on one high performing root server, hetzner provides.

I hope this questions aren't off topic, but I think they are basic questions when using this repo (at least for me)

**Reply by phaer** [reply]
Hi @nupplaphil, great to have you!

> but it wasn't as professional as this solution ;-)

Your smilie implies that we have the same understanding here, but just to be clear: This project isn't really professional in the sense of "reliable" yet. We went to quite a few breaking changes recently and while we are working on release management right now, it will probably take some time before I would recommend to run production workloads on it. We are always looking for more hands to help! 

> since i'm currently heavily using the hetzner root server

This project does not support Hetzners root servers atm, only Hetzner Cloud machines. I would be very interested in supporting root servers as well in the future, but I am not aware of a terraform provider which can provision them yet.

> How are you managing these three parts at your setup? Using three different root paths? Combined at one structure based on this repo?

I personally just include the `kube-hetzner` repo into my own local terraform module where I add things such as Cloudflare DNS records. Everything else (cert-manager, postgres, apps, etc) is then deployed via kustomize. So `kube-hetzner` gets me to the point where i can just run `kubectl` locally and the rest is done via https://kustomize.io/ . This is currently just a playground & test environment for me, I am not using it in production.

> What about the database? Do you recommend using Cassandra with your kubernetes solution or rely on one high performing root server, hetzner provides.

I'd say that heavily depends on your use case. I think it's worth noting, that disk I/O speed isn't the strong-suit of Hetzner Cloud in general and hosting a highly-available database on Kubernetes, especially if storage is backed by external volumes as with our hetzner CSI, comes with quite a lot of pitfalls.

> I hope this questions aren't off topic, but I think they are basic questions when using this repo (at least for me)

Not at all, that's exactly what discussions in this repo are for: keeping issues clean while still providing an opportunity to talk about more open-ended questions like this. Depending on participation, it might make sense to discuss the different topics in separate threads lat er on :)

**Reply by mysticaltech** [reply]
Hey @nupplaphil, thanks for your interest! @phaer pretty much summed up everything. But I would just like to add a few things. Normally we are past the big breaking chances, so personally, I will start using it in production for personal projects.

As for extra things, definitely using kustomize seems great, and even better in combination with something like [ArgoCD](https://argoproj.github.io/cd/), that is the direction I want to go myself (am also a Kubernetes newbie).

As for DBs, I love [ArangoDB](https://www.arangodb.com/) personally, as it's multi-model and runs great on Kubernetes. But many other DBs do have simple Kubernetes solutions now. Just use Helm to install them.

I love the helm packages from Bitnami (part on VMware), always up to date and you can customize everything in the helm values file. It also makes upgrades so easy! 

Last but not least, about disk I/O, the SSD of the Hetzner volumes sounds great to me, but... Am probably not as advanced on the subject as @phaer :)

About setting up the extras, do not hesitate to open more discussion about each software you need here, am very curious myself. I now use only external-dns on top of what kube-hetzner brings (also installed via helm) and looking into setting up ArgoCD soon. The good thing with Kubernetes is that the sky is the limit... In one command you can have any good software under the sun running in your cluster! 

Anyways, just try it if you feel like it, and you will probably get hooked! :)

**Reply by mysticaltech** [reply]
All that said, a member of our team knows quite a bit about DBs! @mnencia What do you think about running let's say Postgres on Hetzner?

@nupplaphil asked above:
> What about the database? Do you recommend using Cassandra with your kubernetes solution or rely on one high performing root server, hetzner provides.

**Reply by mnencia** [reply]
@phaer is right about the pitfalls of running a DB on Kubernetes, but things are rapidly changing as today there is a lot of interest in running DBs on Kubernetes. I suggest you look at [Data on Kubernetes](https://dok.community/) community and join the slack channel. 

If your database is not so big and your workload doesn't require a lot of IOPS I think that is only a matter of choosing the right operator and letting it handle all the details of managing your DB of choice.

**Reply by mnencia** [reply]
> All that said, a member of our team knows quite a bit about DBs! @mnencia What do you think about running let's say Postgres on Hetzner?

Postgres runs well even on a raspberry 😄 , my primary advice is, whatever you want to do with Postgres on any kind of setup, the first step should be to benchmark it. Maybe you can take a look at this tool https://github.com/EnterpriseDB/cnp-bench

**DISCLAIMER: I work for EnterpriseDB, and Cloud Native Postgres is one product that I develop**

**Reply by mysticaltech** [reply]
Thanks for sharing @mnencia !!

**Comment by MihaiNueleanu** [comment]
This is an absolutely amazing project. I'm shocked at how easy it is to provision a reasonably sturdy and fairly intricate kubernetes cluster on hetzner! Such a time saver!

**Reply by mysticaltech** [reply]
Thank you @MihaiNueleanu! 🙏
